core/commands/filestore.go

```diff
@@ -37,58 +49,92 @@ The output is:
 <hash> <size> <path> <offset>
 `,
 	},
-	Arguments: []cmds.Argument{
-		cmds.StringArg("obj", false, true, "Cid of objects to list."),
+	Arguments: []cmdsutil.Argument{
+		cmdsutil.StringArg("obj", false, true, "Cid of objects to list."),
 	},
-	Run: func(req cmds.Request, res cmds.Response) {
-		_, fs, err := getFilestore(req)
+	Run: func(req cmds.Request, re cmds.ResponseEmitter) {
+		_, fs, err := getFilestore(req.InvocContext())
 		if err != nil {
-			res.SetError(err, cmds.ErrNormal)
+			re.SetError(err, cmdsutil.ErrNormal)
 			return
 		}
 		args := req.Arguments()
 		if len(args) > 0 {
-			out := perKeyActionToChan(args, func(c *cid.Cid) *filestore.ListRes {
+			out := perKeyActionToChan(req.Context(), args, func(c *cid.Cid) *filestore.ListRes {
 				return filestore.List(fs, c)
-			}, req.Context())
-			res.SetOutput(out)
+			})
+
+			err = re.Emit(out)
 		} else {
 			next, err := filestore.ListAll(fs)
 			if err != nil {
-				res.SetError(err, cmds.ErrNormal)
+				re.SetError(err, cmdsutil.ErrNormal)
 				return
 			}
-			out := listResToChan(next, req.Context())
-			res.SetOutput(out)
+
+			out := listResToChan(req.Context(), next)
+			err = re.Emit(out)
+			if err != nil {
+				log.Error(err)
+			}
 		}
 	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
-			outChan, ok := res.Output().(<-chan interface{})
-			if !ok {
-				return nil, u.ErrCast()
-			}
-			errors := false
-			for r0 := range outChan {
-				r := r0.(*filestore.ListRes)
-				if r.ErrorMsg != "" {
-					errors = true
-					fmt.Fprintf(res.Stderr(), "%s\n", r.ErrorMsg)
+	PostRun: cmds.PostRunMap{
+		cmds.CLI: func(req cmds.Request, re cmds.ResponseEmitter) cmds.ResponseEmitter {
+			reNext, res := cmds.NewChanResponsePair(req)
+
+			go func() {
+				defer re.Close()
+
+				var (
+					v      interface{}
```

Comment:
I need err and errors at the bottom. And when doing `v, err :=` in an inner block it should refer to a different `err`. And even if it doesn't this way it's more obvious.